synapse/federation/sender/__init__.py

```diff
@@ -353,12 +353,16 @@ async def _process_event_queue_loop(self) -> None:
                 last_token = await self.store.get_federation_out_pos("events")
                 (
                     next_token,
-                    events,
                     event_to_received_ts,
-                ) = await self.store.get_all_new_events_stream(
+                ) = await self.store.get_all_new_event_ids_stream(
                     last_token, self._last_poked_id, limit=100
                 )
 
+                event_entries = await self.store.get_events_from_cache_or_db(
+                    event_to_received_ts.keys()
+                )
+                events = [e.event for e in event_entries.values()]
```

Comment:
I think the code below makes assumptions about the ordering of `events`, and I don't think `get_events_from_cache_or_db` makes any promises about the ordering of its result; so I think you need to sort this to match the same order as `event_to_received_ts`.

It might be better, instead, to leave it as a dict, and change the loop that populates `events_by_room` below to iterate over `event_to_received_ts`.

Reply:
I iterated over the `event_entries` dict instead as I found that `event_to_received_ts` can contain event ids that are not returned by `get_events_from_cache_or_db` (due to being rejected), and it seemed cleaner than iterating over `event_to_received_ts` and checking if the current event_id is present in `event_entries`, then accessing it if it is.